packages/blocks/src/paragraph-block/paragraph-block.ts

```diff
@@ -73,36 +74,10 @@ export class ParagraphBlockComponent extends BlockComponent<
   @query('.affine-paragraph-placeholder')
   private accessor _placeholderContainer: HTMLElement | null = null;
 
-  private _currentTextSelection: TextSelection | undefined = undefined;
+  private _displayPlaceholder!: ReadonlySignal<boolean>;
 
   override accessor blockContainerStyles = { margin: '10px 0' };
 
-  //TODO(@Flrande) wrap placeholder in `rich-text` or inline-editor to make it more developer-friendly
-  private _updatePlaceholder = () => {
```

Comment:
So the benefit of using signal here should be removing more imperative `_updateSomething` methods across the code base?

Also looks great to avoid `updateComplete`! Not sure if this is something can be refactored easily, or requires case by case debugging in time sequence?

Reply:
With the selection signal. Blocks can compute the correct state based on the selection and has mainly 3 benefits:
1. Developers won't need to subscribe to selection state. The compute based pattern will make more straight forward. 
2. Developers won't need to cared about component update triggered by changes (@lit-labs/preact-signal will handle that)
3. The performance will be better because compute is lazy.

However, this cannot be auto refactored by something like a codemod. Although it's not hard to refactor, it still needs us to refactor the code progressively.